client/packages/common/src/ui/forms/JsonForms/components/IdGenerator.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { useCallback, useMemo } from 'react';
+import {
+  composePaths,
+  ControlProps,
+  rankWith,
+  uiTypeIs,
+} from '@jsonforms/core';
+import { withJsonFormsControlProps } from '@jsonforms/react';
+import { BasicTextInput, Box } from '@openmsupply-client/common';
+import { Button, FormLabel } from '@mui/material';
+import {
+  FORM_LABEL_COLUMN_WIDTH,
+  FORM_INPUT_COLUMN_WIDTH,
+} from '../styleConstants';
+
+export const idGeneratorTester = rankWith(4, uiTypeIs('IdGeneratorControl'));
+
+type GeneratorOptions = {
+  targetField: string;
+  parts: Part[];
+};
+
+type Part = {
+  field: string | string[];
+  firstNChars?: number;
+  lastNChars?: number;
+  toUpperCase?: boolean;
+  enumMapping?: Record<string, string>;
+};
+
+const extractField = (
+  data: any,
+  field: string | string[]
+): unknown | undefined => {
+  const fields = typeof field === 'string' ? [field] : field;
+  return fields.reduce((prev, field) => {
+    return prev?.[field];
+  }, data);
+};
+
+const validateFields = (
+  options: GeneratorOptions,
+  data: any
+): string | undefined => {
+  for (const part of options.parts ?? []) {
+    const field = extractField(data, part.field);
+    const fieldName =
+      typeof part.field === 'string' ? part.field : part.field.join('.');
+    if (field === undefined || typeof field !== 'string' || field === '') {
+      return `Missing required field: ${fieldName}`;
+    }
+  }
+};
+
+const generateId = (options: GeneratorOptions, data: any): string => {
```

Comment:
:nit this might work, and make ts happier

```suggestion
const generateId = (options: GeneratorOptions, data: Record<string, unknown>): string => {
```